contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,381 @@
+     * be closed down and disallowed to all. The caller must be the default admin.
+     */
+    function setContractClosed(address target) public virtual onlyDefaultAdmin {
+        bytes32 group = _GROUP_CLOSED;
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Returns true if the team has already been created via {createTeam}.
+     */
+    function _teamExists(uint8 team) internal virtual returns (bool) {
+        return _getTeam(_createdTeams, team);
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_grantRole} that keeps track of user team bitmaps.
+     */
+    function _grantRole(bytes32 role, address user) internal virtual override {
+        super._grantRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            _userTeams[user] = _setTeam(_userTeams[user], team, true);
+        }
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_revokeRole} that keeps track of user team bitmaps.
+     */
+    function _revokeRole(bytes32 role, address user) internal virtual override {
+        super._revokeRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            require(team != _TEAM_PUBLIC);
+            _userTeams[user] = _setTeam(_userTeams[user], team, false);
+        }
+    }
+
+    /**
+     * @dev Returns the {AccessControl} role id that corresponds to a team.
+     *
+     * This role id starts with the ASCII characters `team:`, followed by zeroes, and ends with the single byte
+     * corresponding to the team number.
+     */
+    function _encodeTeamRole(uint8 team) internal virtual view returns (bytes32) {
+        return bytes32("team:") | bytes32(uint256(team));
+    }
+
+    /**
+     * @dev Decodes a role id into a team, if it is a role id of the kind returned by {_encodeTeamRole}.
+     */
+    function _decodeTeamRole(bytes32 role) internal virtual returns (bool, uint8) {
+        bytes32 tagMask = ~bytes32(uint256(0xff));
+        bytes32 tag = role & tagMask;
+        uint8 team  = uint8(role[31]);
+        return (tag == bytes32("team:"), team);
+    }
+
+    /**
+     * @dev Returns the isolate group id for a target contract.
+     *
+     * The group id consists of the ASCII characters `group:isolate:` followed by the contract address bytes.
+     */
+    function _encodeIsolateGroup(address target) internal virtual pure returns (bytes32) {
+        return _GROUP_ISOLATE_PREFIX | (bytes20(target) >> _GROUP_ISOLATE_PREFIX.length);
+    }
+
+    /**
+     * @dev Returns the isolate group id for a target contract.
+     *
+     * The group id consists of the ASCII characters `group:custom:` followed by the group name.
+     */
+    function _encodeCustomGroup(string calldata name) internal virtual pure returns (bytes32) {
```

Comment:
Can we restrict `name` length? I think it's not clear that the last part gets trimmed, and dependent frontends (or even other contracts) may make mistakes

Reply:
The length is restricted in `_encodeCustomGroup`.